connectors/sources/gmail.py

```diff
@@ -0,0 +1,354 @@
+                "required": True,
+            },
+            "customer_id": {
+                "display": "text",
+                "label": "Google customer id",
+                "order": 3,
+                "tooltip": "Google admin console -> Account -> Settings -> Customer Id",
+                "type": "str",
+                "value": "",
+                "required": True,
+            },
+            "use_document_level_security": {
+                "display": "toggle",
+                "label": "Enable document level security",
+                "order": 4,
+                "tooltip": "Document level security ensures identities and permissions set in GMail are maintained in Elasticsearch. This enables you to restrict and personalize read-access users have to documents in this index. Access control syncs ensure this metadata is kept up to date in your Elasticsearch documents.",
+                "type": "bool",
+                "value": True,
+            },
+        }
+
+    async def validate_config(self):
+        """Validates whether user inputs are valid or not for configuration field.
+        Raises:
+            Exception: The format of service account json is invalid.
+        """
+        self.configuration.check_valid()
+
+        try:
+            json.loads(self.configuration["service_account_credentials"])
+        except ValueError as e:
+            raise ConfigurableFieldValueError(
+                f"Google Drive service account is not a valid JSON. Exception: {e}"
+            ) from e
+
+    def advanced_rules_validators(self):
+        return [GMailAdvancedRulesValidator()]
+
+    def _set_internal_logger(self):
+        self._google_directory_client.set_logger(self._logger)
+
+    @cached_property
+    def _service_account_credentials(self):
+        service_account_credentials = json.loads(
+            self.configuration["service_account_credentials"]
+        )
+        return service_account_credentials
+
+    @cached_property
+    def _google_directory_client(self):
+        return GoogleDirectoryClient(
+            json_credentials=self._service_account_credentials,
+            customer_id=self.configuration["customer_id"],
+        )
+
+    def _gmail_client(self, subject):
+        """Instantiates a GMail client for a corresponding subject.
+        Args:
+            subject (str): Email address for the subject.
+
+        Returns:
+            gmail_client (GMailClient): GMail client for the corresponding subject.
+        """
+        gmail_client = GMailClient(
+            json_credentials=self._service_account_credentials,
+            customer_id=self.configuration["customer_id"],
+            subject=subject,
+        )
+        gmail_client.set_logger(self._logger)
+        return gmail_client
+
+    async def ping(self):
+        for service_name, client in [
+            ("GMail", self._gmail_client(self.configuration["subject"])),
+            ("Google directory", self._google_directory_client),
```

Comment:
Intuitively, should directory be above GMail?

I.g. if pinging directory fails, is gmail even gonna be accessible?

Reply:
I would assume that both systems are fairly independent from each other. I can still move Google Directory up, if you think it's a better approach